sqlite.zig

```diff
@@ -590,6 +590,150 @@ pub const Db = struct {
     pub fn savepoint(self: *Self, name: []const u8) Savepoint.InitError!Savepoint {
         return Savepoint.init(self, name);
     }
+
+    pub const CreateFunctionFlag = struct {
+        deterministic: bool = true,
+        direct_only: bool = true,
+    };
+
+    /// Creates a scalar SQLite function with the given name.
+    ///
+    /// When the SQLite function is called in a statement, `func` will be called with the input arguments.
+    /// Each SQLite argument is converted to a Zig value according to the following rules:
+    /// * TEXT values can be either sqlite.Text or []const u8
+    /// * BLOB values can be either sqlite.Blob or []const u8
+    /// * INTEGER values can be any Zig integer
+    /// * REAL values can be any Zig float
+    ///
+    /// TODO(vincent): check for integer and float sizes.
+    ///
+    /// The return type of the function is converted to a SQLite value according to the same rules but reversed.
+    ///
+    pub fn createScalarFunction(self: *Self, func_name: [:0]const u8, comptime func: anytype, comptime create_flags: CreateFunctionFlag) !void {
+        const Type = @TypeOf(func);
+
+        const fn_info = switch (@typeInfo(Type)) {
+            .Fn => |fn_info| fn_info,
+            else => @compileError("expecting a function"),
+        };
+        if (fn_info.is_generic) @compileError("function can't be generic");
+        if (fn_info.is_var_args) @compileError("function can't be variadic");
+
+        const ArgTuple = std.meta.ArgsTuple(Type);
+
+        var flags: c_int = c.SQLITE_UTF8;
+        if (create_flags.deterministic) {
+            flags |= c.SQLITE_DETERMINISTIC;
+        }
+        if (create_flags.direct_only) {
+            flags |= c.SQLITE_DIRECTONLY;
+        }
+
+        const result = c.sqlite3_create_function_v2(
+            self.db,
+            func_name,
+            fn_info.args.len,
+            flags,
+            null,
+            struct {
+                fn sliceFromValue(sqlite_value: *c.sqlite3_value) []const u8 {
+                    const size = @intCast(usize, c.sqlite3_value_bytes(sqlite_value));
+
+                    const value = c.sqlite3_value_text(sqlite_value);
+                    debug.assert(value != null); // TODO(vincent): how do we handle this properly ?
```

Comment:
This function (and `bindValue`; and hence also `xFunc`) should probably return an error set that includes e.g. `OutOfMemory`

Reply:
I don't think I can do that, `xFunc` can't return anything